### ISSUE TITLE
csharp_prefer_simple_default_expression has faulty heuristic

### ISSUE THREAD

**Issue by kentcb** [dotnet/roslyn#22184]
**Version Used**: 

VS Enterprise 2017 15.3.4

**Steps to Reproduce**:

1. Add `csharp_prefer_simple_default_expression` to your .editorconfig with a value of `true:error`
2. Add the statement `object i = default(int)` to your code.

**Expected Behavior**:

No error should be triggered, since the target type (`object`) is more general than the type specified to `default`.

**Actual Behavior**:

An error is triggered, suggesting the statement be simplified to:
```C#
object i = default;
```
Which obviously has different semantics.

This problem is particularly pertinent to the declaration of dependency or bindable properties:
```C#
public static readonly BindableProperty SomeProperty = BindableProperty.Create(
    nameof(Some),
    typeof(Me),
    defaultValue: default(int));
```
This triggers the rule when it should not, since `defaultValue` is of type `object`.

Here is a repro: [EditorConfigRepro.zip](https://github.com/dotnet/roslyn/files/1312593/EditorConfigRepro.zip)

**Comment by sharwell** [comment]
⚠️ When fixing this issue, please include the test case from #22748.

**Comment by jnm2** [comment]
This is bugging me pretty regularly.

**Comment by CyrusNajmabadi** [comment]
Strange.  We use a speculation analyzer here to check if ReplacementChangesSemantics.  I'm curious why that's not returning true for cases like this.  Certainly the type changing from 'Int32' to 'Object' should flag this as a semantics change...

**Comment by d1mnewz** [comment]
related [csharplang/#1080](https://github.com/dotnet/csharplang/issues/1080)

**Comment by CyrusNajmabadi** [comment]
This does not repro for me.  ```int i = default(int)``` offers to change this to ```int i = default```.  But ```object o = default(int)``` does not offer the change.

**Comment by kentcb** [comment]
@CyrusNajmabadi What version of VS?

![image](https://user-images.githubusercontent.com/1901832/32809603-53d53e6c-c9e7-11e7-98ca-5608a1ec9860.png)

**Comment by CyrusNajmabadi** [comment]
15.5p3:

Available when valid:

![image](https://user-images.githubusercontent.com/4564579/32809884-89aa3d98-c94d-11e7-9fb4-1eb7e47527ea.png)

But not when invalid:

![image](https://user-images.githubusercontent.com/4564579/32809895-93adb996-c94d-11e7-928b-9a8170b6944b.png)

**Comment by jcouv** [comment]
I'll go ahead and close since it was fixed already.
@kentcb Feel free to re-open if you still run into this with 15.5 (or later). Thanks
